contracts/TwabRewards.sol

```diff
@@ -0,0 +1,394 @@
+     */
+    function _getPromotion(uint256 _promotionId) internal view returns (Promotion memory) {
+        return _promotions[_promotionId];
+    }
+
+    /**
+        @notice Get the current epoch id of a promotion.
+        @param _promotionId Id of the promotion to get current epoch for
+        @return Epoch id
+     */
+    function _getCurrentEpochId(uint256 _promotionId) internal view returns (uint256) {
+        Promotion memory _promotion = _getPromotion(_promotionId);
+        uint256 _numberOfEpochs = _promotion.numberOfEpochs;
+
+        uint256 _promotionEndTimestamp = _promotion.startTimestamp +
+            (_promotion.epochDuration * _numberOfEpochs);
+        uint256 _currentEpochId = (_numberOfEpochs / block.timestamp) * _promotionEndTimestamp;
+
+        return _currentEpochId;
+    }
+
+    /**
+        @notice Get reward amount for a specific user.
+        @dev Rewards can only be claimed once the epoch is over.
+        @param _user User to get reward amount for
+        @param _promotionId Promotion id from which the epoch is
+        @param _epochId Epoch id to get reward amount for
+        @return Reward amount
+     */
+    function _calculateRewardAmount(
+        address _user,
+        uint256 _promotionId,
+        uint256 _epochId
+    ) internal view returns (uint256) {
+        Promotion memory _promotion = _getPromotion(_promotionId);
+
+        uint256 _epochDuration = _promotion.epochDuration;
+        uint256 _epochStartTimestamp = _promotion.startTimestamp + (_epochDuration * _epochId);
+        uint256 _epochEndTimestamp = _epochStartTimestamp + _epochDuration;
+
+        require(
+            block.timestamp >= _epochStartTimestamp && block.timestamp <= _epochEndTimestamp,
+            "TwabRewards/epoch-not-over"
+        );
+
+        ITicket _ticket = ITicket(_promotion.ticket);
+
+        uint256 _averageBalance = _ticket.getAverageBalanceBetween(
+            _user,
+            uint64(_epochStartTimestamp),
+            uint64(_epochEndTimestamp)
+        );
+
+        uint64[] memory _epochStartTimestamps = new uint64[](1);
+        _epochStartTimestamps[0] = uint64(_epochStartTimestamp);
+
+        uint64[] memory _epochEndTimestamps = new uint64[](1);
+        _epochEndTimestamps[0] = uint64(_epochEndTimestamp);
+
+        uint256[] memory _averageTotalSupplies = _ticket.getAverageTotalSuppliesBetween(
+            _epochStartTimestamps,
+            _epochEndTimestamps
+        );
+
+        return (_promotion.tokensPerEpoch * _averageBalance) / _averageTotalSupplies[0];
+    }
+
+    /**
+        @notice Get the total amount of tokens left to be rewarded.
+        @param _promotionId Promotion id to get the total amount of tokens left to be rewarded for
+        @return Amount of tokens left to be rewarded
+     */
+    function _getRemainingRewards(uint256 _promotionId) internal view returns (uint256) {
+        return IERC20(_getPromotion(_promotionId).token).balanceOf(address(this));
```

Comment:
Whoa!

Just realized this will clear out ALL tokens!  Yikes.

If a promotion is cancelled the number of tokens remaining is the number of remaining epochs * token amount per epoch.

It should include the current epoch as well.

Reply:
Done: https://github.com/pooltogether/v4-periphery/pull/8/commits/2360a14e43f7ab1c1c6b3c3a3a85a62ddee58593